OTP documentation (#558)
vapor/docs: 4.0/docs/crypto.md

Login with Bcrypt passwords can be implemented by first fetching the user's password digest from the database by email or username. The known digest can then be verified against the supplied plaintext password.

## OTP

Vapor supports both HOTP and TOTP one-time passwords. The OTP's work with the SHA-1, SHA-256 and SHA-512 hash functions and can provide six, seven or eight digits of output.

#### HOTP

> ```swift

Reviewer: A brief description of what HOTP is (and maybe how it differs from TOTP?) could be helpful here before the code snippet.

let key = SymmetricKey(size: .bits128)
let hotp = HOTP(key: key, digest: .sha256, digits: .six)
let code = hotp.generate(counter: 25)

// Or using the static generate function
HOTP.generate(key: key, digest: .sha256, digits: .six, counter: 25)
```

#### TOTP

> ```swift

Reviewer: Same here.

Login with Bcrypt passwords can be implemented by first fetching the user's password digest from the database by email or username. The known digest can then be verified against the supplied plaintext password.

## OTP

> Vapor supports both HOTP and TOTP one-time passwords. The OTP's work with the SHA-1, SHA-256 and SHA-512 hash functions and can provide six, seven or eight digits of output. An OTP is able to provide authentication by generating a human-readable password, which must only be used once. To do so, parties first agree on a symmetric key (which must be kept private at all times).

Reviewer 2: - In the second sentence, replace `The OTP's` with `OTPs` - there is more than one OTP (plural, not possessive), and the article is erroneously specific in this context.
- In the second sentence, add a comma immediately after `SHA-256` (Oxford comma).
- Also add a comma immediately after `seven` (Oxford comma).
- Suggest rephrasing the third sentence as "An OTP provides authentication by generating a single-use human-readable password." (active voice, simplified phrasing)
- Suggest removing the parenthesis on the fourth sentence and adding clarification, e.g. `"... agree on a symmetric key, which must be kept private at all times to maintain the security of the generated passwords." (high-importance subclause, emphasis on security requirement)

The resulting paragraph would be:
```suggestion
Vapor supports both HOTP and TOTP one-time passwords. OTPs work with the SHA-1, SHA-256, and SHA-512 hash functions and can provide six, seven, or eight digits of output. An OTP provides authentication by generating a single-use human-readable password. To do so, parties first agree on a symmetric key, which must be kept private at all times to maintain the security of the generated passwords.
```

#### HOTP

> The HOTP is an OTP based on HMAC's. In addition to the symmetric key, both parties also agree on a counter, which is a number, providing uniqueness for the password. After each generation attempt, the counter is increased.

Reviewer 2: - Remove "The" (excessive use of article)
- Replace `HMAC's` with `an HMAC digest` (plural, not possessive, and use a more descriptive phrasing).
- Remove the comma between "number" and "providing" (unnecessary subclause division).
Result:
```suggestion
HOTP is an OTP based on an HMAC digest. In addition to the symmetric key, both parties also agree on a counter, which is a number providing uniqueness for the password. After each generation attempt, the counter is increased.
```

Author: Omit the word digest, HMAC is not a digest algorithm, but a MAC algorithm.

Reviewer 2: Good catch! I was thinking of the underlying hash; technically HMAC is a signature.

```swift
let key = SymmetricKey(size: .bits128)
let hotp = HOTP(key: key, digest: .sha256, digits: .six)
let code = hotp.generate(counter: 25)

// Or using the static generate function
HOTP.generate(key: key, digest: .sha256, digits: .six, counter: 25)
```

#### TOTP

> A TOTP is an addition to the HOTP. It works the same, but it uses time instead of a counter to generate uniqueness. To compensate for minor skews introduced by unsynchronized clocks, network latency and user delay, a TOTP is generated to be valid over a specified time interval.

Reviewer 2: Suggest rephrasing the entire paragraph:
```suggestion
A TOTP is a time-based variation of the HOTP. It works mostly the same, but instead of a simple counter, the current time is used to generate uniqueness. To compensate for the inevitable skew introduced by unsynchronized clocks, network latency, user delay, and other confounding factors, a generated TOTP code remains valid over a specified time interval (most commonly, 30 seconds).
```

```swift
let key = SymmetricKey(size: .bits128)
let totp = TOTP(key: key, digest: .sha256, digits: .six, interval: 60)
let code = totp.generate(time: Date())

// Or using the static generate function
TOTP.generate(key: key, digest: .sha256, digits: .six, interval: 60, time: Date())
```

#### Range

> OTP's are very useful for giving some leeway in validation and out of sync counters. Both OTP implementations have the ability to generate the OTP for an error margin.

Reviewer 2: - Replace `OTP's` with `OTPs` (possessive -> plural)
- Replace `giving some` with `providing` (phrasing)
- Replace "the OTP for an error margin" with "an OTP with a margin for error" (phrasing)
Result:
```suggestion
OTPs are very useful for providing leeway in validation and out of sync counters. Both OTP implementations have the ability to generate an OTP with a margin for error.
```

```swift
let key = SymmetricKey(size: .bits128)
let hotp = HOTP(key: key, digest: .sha256, digits: .six)

// Generate a window of correct counters
let codes = hotp.generate(counter: 25, range: 2)
```

> The example above allows for a margin of 2, which means the HOTP will be calculated for the counter value `23 ... 27` and all of these codes will be returned. 

Reviewer 2: - Replace `value` with `values` (more than one)
- Add a comma after the range (run-on clause)
```suggestion
The example above allows for a margin of 2, which means the HOTP will be calculated for the counter values `23 ... 27`, and all of these codes will be returned. 
```

!!! warning

>     Note: giving a higher margin will reduce the security as an attacker has more time or freedom to perform attacks.

Reviewer 2: Rephrase:
```suggestion
    Note: The larger the error margin used, the more time and freedom an attacker has to act, decreasing the security of the algorithm.
```

```

#### TOTP

> A TOTP is an addition to the HOTP. It works the same, but it uses time instead of a counter to generate uniqueness. Due to latency problems, both networks but also humans related, the TOTP is generated over a time interval.

Reviewer 3: > Due to latency problems, both networks but also humans related, the TOTP is generated over a time interval.

This feels a bit clunky to me – perhaps something along the lines of:

To compensate for minor skews introduced by unsynchronized clocks, network latency and user delay, a TOTP is generated to be valid over a specified time interval.

Reviewer: 👍
